[OpenTelemetry.Instrumentation.AWSLambda]: update README according to the recent code changes. (#1234)
open-telemetry/opentelemetry-dotnet-contrib: src/OpenTelemetry.Instrumentation.AWSLambda/README.md

```csharp
TracerProvider tracerProvider = Sdk.CreateTracerProviderBuilder()
        // add other instrumentations
        .AddAWSLambdaConfigurations(options => options.DisableAwsXRayContextExtraction = true)
        .Build();
```

### AWSLambdaInstrumentationOptions

`AWSLambdaInstrumentationOptions` contains various properties to configure

> AWS lambda instrumentation.

Reviewer: ```suggestion
AWS lambda instrumentation:
```

* [`DisableAwsXRayContextExtraction`](/src/OpenTelemetry.Instrumentation.AWSLambda/AWSLambdaInstrumentationOptions.cs)
* [`SetParentFromBatch`](/src/OpenTelemetry.Instrumentation.AWSLambda/AWSLambdaInstrumentationOptions.cs)

## Instrumentation

> `AWSLambdaWrapper` class contains tracing methods covering different types of

Reviewer: ```suggestion
`AWSLambdaWrapper` contains tracing methods covering different types of
```

or

```suggestion
The `AWSLambdaWrapper` class contains tracing methods covering different types of
```

> function handler method signatures. `AWSLambdaWrapper.Trace()` and
> `AWSLambdaWrapper.TraceAsync()` methods are used for wrapping synchronous

Reviewer: ```suggestion
function handler method signatures. `AWSLambdaWrapper.Trace()` and
`AWSLambdaWrapper.TraceAsync()` are used for wrapping synchronous
```

or

```suggestion
function handler method signatures. The methods `AWSLambdaWrapper.Trace()` and
`AWSLambdaWrapper.TraceAsync()` are used for wrapping synchronous
```

and asynchronous function handlers respectively. The `ActivityContext parentContext`
parameter is optional and used to pass a custom parent context. If the parent
is not passed explicitly then it's either extracted from the
input parameter or uses AWS X-Ray headers if AWS X-Ray context extraction is
enabled (see configuration property `DisableAwsXRayContextExtraction`).
The sequence of the parent extraction:
`explicit parent` -> `parent from input parameter` -> `AWS X-Ray headers` -> `default`
The parent extraction is supported for the input types listed in the table below.
Although the `SQSEvent` and `SNSEvent` consist of multiple

> records (`SQSMessage` and `SNSRecord`) it's recommended to pass the full event.

Author: Better to remove it.

> | Package | Types |

Reviewer: I wonder if there is any  value in listing the Package? Instead, it would be good to mention briefly where it is extracted from (e.g. for SQSEvent: Message attributes of last message; API Gateway events: HTTP headers)

Author: I replaced the 'Package' column by the 'Parent extraction source'.

|---------|-------|
| `Amazon.Lambda.APIGatewayEvents` | `APIGatewayProxyRequest, APIGatewayHttpApiV2ProxyRequest` |
| `Amazon.Lambda.SQSEvents` | `SQSEvent` |
| `Amazon.Lambda.SNSEvents` | `SNSEvent` |

### Lambda Function

> 1. Create a wrapper function with the same signature as the original Lambda function.

Reviewer: I think this is not true in all cases, as AFAIK we require the context parameter now.

```suggestion
1. Create a wrapper function with the same signature as the original Lambda function but an added ILambdaContext parameter if it was not already present.
```

> Call `AWSLambdaWrapper.Trace()` or `AWSLambdaWrapper.TraceAsync()` API and pass
> `TracerProvider`, original Lambda function
> and its inputs as parameters.

Reviewer: ```suggestion
Call `AWSLambdaWrapper.Trace()` or `AWSLambdaWrapper.TraceAsync()` API and pass
`TracerProvider`, original Lambda function and its parameters.
```

```csharp
TracerProvider tracerProvider = Sdk.CreateTracerProviderBuilder()
        // add other instrumentations
        .AddAWSLambdaConfigurations(options => options.DisableAwsXRayContextExtraction = true)
        .Build();
```

### AWSLambdaInstrumentationOptions

`AWSLambdaInstrumentationOptions` contains various properties to configure
AWS lambda instrumentation.

> - [`DisableAwsXRayContextExtraction`](/AWSLambdaInstrumentationOptions.cs#L25)

Reviewer: Linking to a particular line seems dangerous, one would need to remember to update the readme for any change in this file. It will surely get out of sync. I would just link the whole file once, and not link each option. The radical approach would be to not even list them but just point to the file for the list, but one could also list them with a brief one-sentence summary.

Author: Right. I replaced both links by the file reference.

- [`SetParentFromBatch`](/src/OpenTelemetry.Instrumentation.AWSLambda/AWSLambdaInstrumentationOptions.cs)

> Boolean value indicating whether the parent Activity should be set when

Reviewer: Is this paragraph intentionally left? Seems to belong to the second point, or should be removed.

Author: No, unintentionally. Removed.

```csharp
TracerProvider tracerProvider = Sdk.CreateTracerProviderBuilder()
        // add other instrumentations

>         .AddAWSLambdaConfigurations(options => options.DisableAwsXRayContextExtraction = true)

Author: Good idea. Added it to the default example.

>         .AddAWSLambdaConfigurations(options => options.DisableAwsXRayContextExtraction = true)

Reviewer: This was not meant as suggestion, but I am wondering why we show that option in the example. If the option would be good as default-on for some reason, probably we should change it in the implementation, otherwise I would recommend leaving the example as it was before with just

```suggestion
        .AddAWSLambdaConfigurations()
```

And maybe show a call with customized options in the section below.

>         .AddAWSLambdaConfigurations(options => options.DisableAwsXRayContextExtraction = true)

Author: But may be before we change implementation it's worth to keep DisableAwsXRayContextExtraction in the example ? Otherwise it may lead to some misunderstandings.

>         .AddAWSLambdaConfigurations(options => options.DisableAwsXRayContextExtraction = true)

Reviewer 2: In general looks good, I am not AWSLamda expert. If `DisableAwsXRayContextExtraction` set to `true` should be default option lets keep it on both examples. Otherwise remove it.

>         .AddAWSLambdaConfigurations(options => options.DisableAwsXRayContextExtraction = true)

Author: Then lets keep it in both examples as is.

        .Build();
```

### AWSLambdaInstrumentationOptions

`AWSLambdaInstrumentationOptions` contains various properties to configure
AWS lambda instrumentation.

> #### `DisableAwsXRayContextExtraction` (optional)

Reviewer: "(optional)" can be omitted, as all options are optional and this is as expected I think.

Author: Removed.

> #### `DisableAwsXRayContextExtraction` (optional)

Author: I added relative links to the xml doc.

Boolean value indicating whether AWS X-Ray context extraction should be disabled.
Default value is false.

#### `SetParentFromBatch` (optional)

Boolean value indicating whether the parent Activity should be set when
a potentially batched event is received where multiple parents are potentially
available (e.g. SQS). If set to true, the parent is set using the last received
record (e.g. last message). Otherwise the parent is not set. In both cases, links
will be created for such events. Currently, the only event type to which this
applies is SQS. Default value is false.

## Instrumentation

`AWSLambdaWrapper` class contains tracing methods covering different types of
function handler method signatures. `AWSLambdaWrapper.Trace()` and

> `AWSLambdaWrapper.TraceAsync()` methods are used for wrapping synchronious
> and asynchronious function handlers respectively. The `ActivityContext parentContext`

Reviewer: ```suggestion
`AWSLambdaWrapper.TraceAsync()` methods are used for wrapping synchronous
and asynchronous function handlers respectively. The `ActivityContext parentContext`
```

> parameter is optional and used to pass the explicitly extracted parent. If the parent

Reviewer: ```suggestion
parameter is optional and used to pass a custom parent context. If the parent
```

> is not passed from the customer's code then it's either extracted from the

Reviewer: ```suggestion
is not passed explicitly then it's either extracted from the
```

input parameter or uses AWS X-Ray headers if AWS X-Ray context extraction is
enabled (see configuration property `DisableAwsXRayContextExtraction`).
The sequence of the parent extraction:
`explicit parent` -> `parent from input parameter` -> `AWS X-Ray headers` -> `default`
The parent extraction is supported for the following input types:

| Package | Types |
|---------|-------|
| `Amazon.Lambda.APIGatewayEvents` | `APIGatewayProxyRequest, APIGatewayHttpApiV2ProxyRequest` |

> | `Amazon.Lambda.SQSEvents` | `SQSEvent, SQSMessage` |

Reviewer: IMHO, we should not document this. Users should always pass the full record.

Author: I removed individual record types and mentioned that users should always pass the full input object.
